### PR TITLE
v0.2

### PR DESCRIPTION
**Pull request by bboynton97**
This PR represents new features coming in version 0.2

- Managing agent tools is now done with AST
- Projects can be init-ed using templates

### CODE REVIEW

File: agentstack/cli/cli.py [hunk 1]
```diff
@@ -16,17 +18,67 @@
...
+        if template[:len(url_start)] == url_start:
+            # template is a url
+            response = requests.get(template)
+            if response.status_code == 200:
+                template_data = response.json()
+            else:
+                print(term_color(f"Failed to fetch template data from {template}. Status code: {response.status_code}", 'red'))
+                sys.exit(1)
+        else:
+            with importlib.resources.path('agentstack.templates.proj_templates', f'{template}.json') as template_path:
```
Comment (tcdent) [inline agentstack/cli/cli.py:42]: I just added it, but `utils.get_package_path` seems to make these interactions a little more robust, especially across versions and avoids the `with` statement.

File: agentstack/cli/cli.py [hunk 2]
```diff
@@ -38,21 +90,16 @@ def init_project_builder(slug_name: Optional[str] = None, skip_wizard: bool = Fa
...
         tools = []
-    else:
-        welcome_message()
-        project_details = ask_project_details(slug_name)
-        welcome_message()
-        framework = ask_framework()
-        design = ask_design()
-        tools = ask_tools()
 
     log.debug(
```
Comment (tcdent) [inline agentstack/cli/cli.py:94]: I ❤️ logging for the realtime visibility into stuff happening in the background. We should think of it like a tool for the users, too.

File: agentstack/generation/agent_generation.py [hunk 3]
```diff
@@ -21,7 +21,7 @@ def generate_agent(
     if not backstory:
         backstory = 'Add your backstory here'
     if not llm:
-        llm = 'Add your llm here with format provider/model'
+        llm = 'openai/gpt-4o'
```
Comment (tcdent) [inline agentstack/generation/agent_generation.py:24]: We could store this in agentstack.json for future reference.

File: agentstack/generation/gen_utils.py [hunk 4]
```diff
@@ -66,3 +71,72 @@ def string_in_file(file_path: str, str_to_match: str) -> bool:
...
+
+def _framework_filename(framework: str, path: str = ''):
+    if framework == 'crewai':
+        return f'{path}src/crew.py'
+
+    print(term_color(f'Unknown framework: {framework}', 'red'))
+    sys.exit(1)
+
+
+class CrewComponentType(str, Enum):
```
Comment (tcdent) [inline agentstack/generation/gen_utils.py:83]: Enum is perfect but whenever I see how often it's referenced I always regret that I didn't make the name shorter. 

`CrewComponent`?

File: agentstack/generation/task_generation.py [hunk 5]
```diff
@@ -87,3 +87,8 @@ def generate_crew_task(
     ]
 
     insert_after_tasks(file_path, code_to_insert)
+
+
+def get_task_names(framework: str, path: str = '') -> List[str]:
```
Comment (tcdent) [inline agentstack/generation/task_generation.py:92]: Should have a conditional for the framework.

Comment (tcdent) [reply agentstack/generation/task_generation.py:92]: Oh, wait, that does get passed down. `get_crew_components` should be crew-specific.

File: agentstack/generation/tool_generation.py [hunk 6]
```diff
@@ -108,26 +115,25 @@ def remove_tool_from_tools_init(tool_data: dict, path: str = ''):
...
+        """
+    modify_agent_tools(framework, tool_data, 'add', agents, path, 'tools')
 
 
 def remove_tool_from_agent_definition(framework: str, tool_data: dict, path: str = ''):
-    filename = _framework_filename(framework, path)
-    with fileinput.input(files=filename, inplace=True) as f:
-        for line in f:
-            print(line.replace(f'{", ".join([f"tools.{tool_name}" for tool_name in tool_data["tools"]])}, ', ''), end='')
+    modify_agent_tools(framework, tool_data, 'remove', None, path, 'tools')
```
Comment (tcdent) [inline agentstack/generation/tool_generation.py:136]: Consider using keyword arguments here since it's not intuitive what each param is doing.

File: agentstack/generation/tool_generation.py [hunk 7]
```diff
@@ -139,3 +145,201 @@ def assert_tool_exists(tool_name: str, tools: dict):
     print(term_color(f'No known agentstack tool: {tool_name}', 'red'))
     sys.exit(1)
 
+
+def _create_tool_attribute(tool_name: str, base_name: str = 'tools') -> ast.Attribute:
```
Comment (tcdent) [inline agentstack/generation/tool_generation.py:149]: AST let's goooooo

File: agentstack/generation/tool_generation.py [hunk 8]
```diff
@@ -139,3 +145,201 @@ def assert_tool_exists(tool_name: str, tools: dict):
...
+        ctx=ast.Load()
+    )
+
+
+def _create_tool_attributes(
+        tool_names: List[str],
+        base_name: str = 'tools'
+) -> List[ast.Attribute]:
+    """Create AST nodes for multiple tool attributes"""
+    return [_create_tool_attribute(name, base_name) for name in tool_names]
```
Comment (tcdent) [inline agentstack/generation/tool_generation.py:174]: I might bring `_create_tool_attribute` into this function if it's not used anywhere else.

File: agentstack/generation/tool_generation.py [hunk 9]
```diff
@@ -139,3 +145,201 @@ def assert_tool_exists(tool_name: str, tools: dict):
...
+        return node
+
+    # Find the Return statement and modify tools
+    for item in node.body:
+        if isinstance(item, ast.Return):
+            agent_call = item.value
+            if isinstance(agent_call, ast.Call):
+                for kw in agent_call.keywords:
+                    if kw.arg == 'tools':
+                        if isinstance(kw.value, ast.List):
```
Comment (tcdent) [inline agentstack/generation/tool_generation.py:285]: > > > > > > It do be like that sometimes 🤠

File: agentstack/generation/tool_generation.py [hunk 10]
```diff
@@ -139,3 +145,201 @@ def assert_tool_exists(tool_name: str, tools: dict):
...
+                            # Replace with new list
+                            kw.value = ast.List(elts=new_tools, ctx=ast.Load())
+
+    return node
+
+
+def modify_agent_tools(
+        framework: str,
+        tool_data: Dict,
+        operation: str,
```
Comment (tcdent) [inline agentstack/generation/tool_generation.py:303]: I see why you're passing the `operation` from the top, since a lot of the logic is shared in searching the tree to get to the point where you're making an addition or deletion.

File: 0.2-changes.md [hunk 11]
```diff
@@ -0,0 +1,85 @@
+# 0.2 Roadmap
```
Comment (tcdent) [inline 0.2-changes.md:1]: I thought you hated documentation; this is fantastic!

File: agentstack/generation/gen_utils.py [hunk 12]
```diff
@@ -66,3 +71,72 @@ def string_in_file(file_path: str, str_to_match: str) -> bool:
...
+    AGENT = "agent"
+    TASK = "task"
+
+
+def get_crew_components(
+        framework: str = 'crewai',
+        component_type: Optional[Union[CrewComponentType, List[CrewComponentType]]] = None,
+        path: str = ''
+) -> dict[str, List[str]]:
+    """
```
Comment (tcdent) [inline agentstack/generation/gen_utils.py:93]: Also your method docstrings are legit

File: agentstack/cli/cli.py (hunk 1 above)
Comment (bboynton97) [reply agentstack/cli/cli.py:42]: i saw that in the PR! once that's in main, ill use it here instead

File: agentstack/cli/cli.py (hunk 2 above)
Comment (bboynton97) [reply agentstack/cli/cli.py:94]: definitely agree. maybe we can make it more accessible too by changing the logging level if they use a `--verbose` flag on any command

File: agentstack/generation/agent_generation.py (hunk 3 above)
Comment (bboynton97) [reply agentstack/generation/agent_generation.py:24]: like which provider/model they prefer?

File: agentstack/generation/gen_utils.py (hunk 4 above)
Comment (bboynton97) [reply agentstack/generation/gen_utils.py:83]: agreed, i can do this

File: 0.2-changes.md (hunk 11 above)
Comment (bboynton97) [reply 0.2-changes.md:1]: lol i've been trying to keep it up to date with my plan as i've been working on it. ill probably delete it before merging. i figured it doesn't hurt to make the plans public and in sync with commits

File: agentstack/generation/gen_utils.py (hunk 12 above)
Comment (bboynton97) [reply agentstack/generation/gen_utils.py:93]: lol ill tell copiliot you said thank you

File: 0.2-changes.md (hunk 11 above)
Comment (bboynton97) [reply 0.2-changes.md:1]: also the docs i hate writing are more the quickstart guides that are more marketing copy than technical documentation

File: agentstack/generation/task_generation.py (hunk 5 above)
Comment (bboynton97) [reply agentstack/generation/task_generation.py:92]: so this is kind of unclear at the moment.

Crew is a very opinionated framework that has two components: agents and tasks

Autogen is extremely unopinionated and isnt as simple as to have design boil down to agents/tasks. I can explain the pros and cons later, but my assumption is when we implment autogen/AG2 next, we'll need to enforce _some_ opinions of our own in order to make it possible to build a project with it using AgentStack. This very well may end up being agent/task components like Crew.

tl;dr: idk what other framework structures will look like yet, so i'm leaving this function ambiguous intentionally

File: agentstack/generation/tool_generation.py (hunk 6 above)
Comment (bboynton97) [reply agentstack/generation/tool_generation.py:135]: done

File: agentstack/generation/tool_generation.py (hunk 10 above)
Comment (bboynton97) [reply agentstack/generation/tool_generation.py:302]: its annoying but it i had to go back and add that in so many layers because it matters at the very end

**Comment by bboynton97** [comment]
okay i think this is finally it 😮‍💨 im not merging anything else into main until this is in lol. that was a lot of merge conflicts

**Comment by tcdent** [comment]
Went back through and don't see anything beyond a couple extra newlines. 

Let's get it in!

**Comment by bboynton97** [comment]
ope yeah that was weird. didnt notice that! it might be a good idea to add black formatter to the repo actually

**Review by tcdent — APPROVED** [review]
